client/src/entrypoints/admin/page-editor.js

```diff
@@ -402,6 +384,122 @@ window.updateFooterSaveWarning = (formDirty, commentsDirty) => {
+  });
+
+  const refreshButton = previewPanel.querySelector(
+    '.preview-panel__refresh-button',
+  );
+  const newTabButton = previewPanel.querySelector(
+    '.preview-panel__size-button--new-tab',
+  );
+  const iframe = previewPanel.querySelector('.preview-panel__iframe');
+  const form = document.querySelector('.edit-form');
+  const previewUrl = previewPanel.dataset.action;
+  const submitAction = document.querySelector('.action-save');
+  const previewModeSelect = document.querySelector(
+    '.preview-panel__mode-select',
+  );
+
+  const setPreviewData = () =>
+    fetch(previewUrl, {
+      method: 'POST',
+      body: new FormData(form),
+    }).then((response) =>
+      response.json().then((data) => {
+        if (data.is_valid) {
+          previewPanel.classList.remove('preview-panel--has-errors');
+        } else {
+          previewPanel.classList.add('preview-panel--has-errors');
+        }
+        iframe.contentWindow.location.reload();
+        return data.is_valid;
+      }),
+    );
+
+  const handlePreview = () =>
+    setPreviewData().catch(() => {
+      // eslint-disable-next-line no-alert
+      alert('Error while sending preview data.');
+    });
+
+  const handlePreviewInNewTab = () => {
+    const previewWindow = window.open('', previewUrl);
+    previewWindow.focus();
+
+    handlePreview().then((success) => {
+      if (success) {
+        const url = new URL(previewUrl, window.location.origin);
+        if (previewModeSelect && previewModeSelect.value) {
+          url.searchParams.set('mode', previewModeSelect.value);
+        }
+        previewWindow.document.location = url.toString();
+      } else {
+        window.focus();
+        previewWindow.close();
+      }
+    });
+  };
+
+  refreshButton.addEventListener('click', handlePreview);
+  newTabButton.addEventListener('click', handlePreviewInNewTab);
+
+  const handlePreviewModeChange = (event) => {
+    const mode = event.target.value;
+    const url = new URL(iframe.src);
+    url.searchParams.set('mode', mode);
+    // Make sure data is up-to-date before changing the preview mode.
+    handlePreview().then(() => (iframe.src = url.toString()));
+  };
+
+  if (previewModeSelect) {
+    previewModeSelect.addEventListener('change', handlePreviewModeChange);
+  }
+
+  // Make sure current preview data in session exists and is up-to-date.
+  setPreviewData();
+
+  if (previewPanel.dataset.autoUpdate === 'true') {
+    // Form data is changed when field values are changed (change event),
+    // and we need to delay setPreviewData when typing to avoid useless extra
+    // AJAX requests (so we postpone setPreviewData when keyup occurs).
+
+    let autoUpdatePreviewDataTimeout = -1;
+    const autoUpdatePreview = () => {
+      clearTimeout(autoUpdatePreviewDataTimeout);
+      autoUpdatePreviewDataTimeout = setTimeout(setPreviewData, 1000);
+    };
+
+    ['change', 'keyup'].forEach((e) =>
+      form.addEventListener(e, autoUpdatePreview),
+    );
```

Comment:
Yeah, I previously added a `MutationObserver` to handle those cases but it seemed overkill (discussed with LB above).

It should be possible to implement that logic, but it will add some complexity. We initially didn't think of including automatic preview updates in this PR (hence the existence of the Refresh button), but I thought maybe it's worth adding a simple implementation as the code already exists in the old preview (see removed lines ~300-350).